compiler/rustc_codegen_ssa/src/back/link.rs

```diff
@@ -1231,13 +1231,13 @@ pub fn linker_and_flavor(sess: &Session) -> (PathBuf, LinkerFlavor) {
                     sess.emit_fatal(errors::LinkerFileStem);
                 });
 
+                // Find and strip any version postfix (e.g. `-12` in `clang-12`)
+                let version_postfix = Regex::new(r"-\d+$").unwrap();
```

Comment:
If we're using a regex anyway, I'd prefer to have a regex for `^clang(-\d+)?$` and replace `stem == "clang"` with a regex match.

Reply:
It's not just clang that can have a version postfix though. I also have `gcc-xx` and `lld-xx` on my system for example.

Reply:
I think we can avoid regexes here and use `"clang" || startwith("clang-") || endswith("-clang")` (and same for gcc), because if it says clang then it's probably clang regardless of digits.

Reply:
For gcc we would have to add `contains("-gcc-")` in addition to startswith and endswith or directly match on `contains("gcc") to also catch variants like `aarch64-linux-gnu-gcc-11` without a regex. If that is the preferred way, then I can adjust the PR.

Reply:
I think it may be better to use the regex after all. `stem.starts_with("lld-")` will not only match for `lld-12` but also for `lld-link`. Of course if we match for `lld-link` first and the for `stem.starts_with("lld-")` it will work fine, but in my opinion this would be much more fragile than using the regex to check for digits at the end.

Reply:
I pushed a commit which does not use regex and instead uses `starts_with` as suggested. For `lld` it additionally checks if the characters after `lld-` are digits in order to not match against `lld-link`. The CI failure is because I used `is_some_and`. Should we proceed with this approach or roll-back to the regex approach?

Reply:
I have no strong opinion.

Reply:
>  because if it says clang then it's probably clang regardless of digits.

Actually, this is problematic because [`clang-cl`](https://clang.llvm.org/docs/UsersManual.html#clang-cl) exists which mimics the `cli` interface of `cl.exe` and is thus definitely not gnu.

Reply:
We don't support using `clang-cl` as a linker, so specifying it in `-Clinker` won't result in anything good regardless of whether the guessing in `fn linker_and_flavor` is successful or not.

(If we ever support it then placing `clang-cl` before `clang` in the if-else chain will solve the problem trivially.)

Reply:
Thanks, that's good to know! Does the same also apply for `cl.exe`, i.e. linking for MSVC is only supported via `link` and `lld-link`?